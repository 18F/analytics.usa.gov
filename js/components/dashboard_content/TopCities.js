import React, { useEffect, useState } from "react";
import PropTypes from "prop-types";

import DataLoader from "../../lib/data_loader";
import Tooltip from "../tooltip/Tooltip";
import FilterSelect from "../select/FilterSelect";
import ConsolidatedBarChart from "../chart/ConsolidatedBarChart";

/**
 * Retrieves the realtime top cities report from the passed data URL and
 * creates a visualization for the count of users visiting sites for the current
 * agency from each city.
 *
 * @param {object} props the properties for the component
 * @param {string} props.dataHrefBase the URL of the base location of the data
 * to be downloaded including the agency path. In production this is proxied and
 * redirected to the S3 bucket URL.
 * @param {number} props.refreshSeconds the number of seconds to wait before
 * refreshing chart data.
 * @returns {import('react').ReactElement} The rendered element
 */
function TopCities({ dataHrefBase, refreshSeconds }) {
  const reportFilters = [
    ["30 Minutes", "top-cities-realtime"],
    ["Yesterday", "top-cities-yesterday"],
    ["7 Days", "top-cities-7-days"],
    ["30 Days", "top-cities-30-days"],
    ["90 Days", "top-cities-90-days"],
    ["Current Year", "top-cities-current-year"],
    ["Current Fiscal Year", "top-cities-current-fiscal-year"],
    ["Previous Year", "top-cities-previous-year"],
    ["Previous Fiscal Year", "top-cities-previous-fiscal-year"],
  ];
  const [currentFilter, setCurrentFilter] = useState(reportFilters[0]);
  const [refreshInterval, setRefreshInterval] = useState(null);
  const [isRealtime, setIsRealtime] = useState(true);
  const [data, setData] = useState(null);

  useEffect(() => {
    const initChart = async () => {
      if (currentFilter) {
        await loadDataAndBuildChart();

        if (refreshInterval) {
          clearInterval(refreshInterval);
        }

        if (isRealtime) {
          // Refresh data every interval.
          setRefreshInterval(
            setInterval(() => {
              loadDataAndBuildChart();
            }, refreshSeconds * 1000),
          );
        }
      }
    };
    initChart().catch(console.error);
  }, [currentFilter, isRealtime]);

  async function loadDataAndBuildChart() {
    let data;

    try {
      data = await DataLoader.loadJSON(
        `${dataHrefBase}/${currentFilter[1]}.json`,
      );
    } catch (e) {
      data = { data: [], totals: {} };
    }

    await setData(data);
  }

  async function filterChangeHandler(fileName) {
    if (!fileName) return;

    const selectedFilter = reportFilters.find((reportFilter) => {
      return reportFilter[1] == fileName;
    });

    await setIsRealtime(selectedFilter == reportFilters[0]);
    await setCurrentFilter(selectedFilter);
  }

  return (
    <>
      <div className="grid-row">
        <div className="chart__title display-flex card:grid-col-12 mobile-lg:grid-col-fill card:flex-justify-center mobile-lg:flex-justify-start card:padding-bottom-105 mobile-lg:padding-bottom-0">
          <a
            href={`${dataHrefBase}/${currentFilter[1]}.csv`}
            aria-label={`${currentFilter[1]}.csv`}
          >
            <svg
              className="usa-icon margin-bottom-neg-05 margin-left-05"
              aria-hidden="true"
              focusable="false"
              role="img"
            >
              <use xlinkHref="/assets/uswds/img/sprite.svg#file_present"></use>
            </svg>
          </a>
          <a href="/definitions#dimension_city">
            <Tooltip
              position="top"
              content="The city from which user activity originates. Location data may be affected by a user's VPN usage."
            >
              Cities
            </Tooltip>
          </a>
        </div>
        <div className="card:grid-col-12 mobile-lg:grid-col-auto">
          <div className="display-flex card:flex-justify-center mobile-lg:flex-justify-end">
            <FilterSelect
              filters={reportFilters}
              defaultFilterValue={reportFilters[0][1] || ""}
              onChange={filterChangeHandler}
              name={"cities chart time filter"}
            />
          </div>
        </div>
      </div>
      {data && (
        <div className="text--capitalize">
          <ConsolidatedBarChart
            data={data}
            chartDataKey={"city"}
            maxItems={11}
          />
        </div>
      )}
    </>
  );
}

TopCities.propTypes = {
  dataHrefBase: PropTypes.string.isRequired,
  refreshSeconds: PropTypes.number.isRequired,
};

export default TopCities;
