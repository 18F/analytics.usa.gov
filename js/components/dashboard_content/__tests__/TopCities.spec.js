import React from "react";
import { render, screen, waitFor } from "@testing-library/react";

import { delay } from "../../../../spec/support/test_utilities";
import DataLoader from "../../../lib/data_loader";
import TopCities from "../TopCities";
import CitiesReportFactory from "../../../../spec/factories/reports/cities";

jest.mock("../../../lib/data_loader", () => ({
  ...jest.requireActual("../../../lib/data_loader"),
  loadRealtimeReportJSON: jest.fn(),
}));

describe("TopCities", () => {
  let component;

  describe("when data is not loaded", () => {
    beforeEach(() => {
      DataLoader.loadRealtimeReportJSON.mockImplementation(() => {
        return Promise.resolve(null);
      });
      component = render(
        <TopCities
          dataHrefBase="http://www.example.com/data/"
          refreshSeconds={30}
        />,
      );
    });

    it("renders a component in loading state", () => {
      expect(component.asFragment()).toMatchSnapshot();
    });
  });

  describe("when data is loaded", () => {
    const data = CitiesReportFactory.build();

    beforeEach(async () => {
      data.totals.by_city = {};
      data.data.forEach((item) => {
        data.totals.by_city[item.city] = item.activeUsers;
      });
      DataLoader.loadRealtimeReportJSON.mockImplementation(() => {
        return Promise.resolve(data);
      });
      component = render(
        <TopCities
          dataHrefBase="http://www.example.com/data/"
          refreshSeconds={30}
        />,
      );
      await waitFor(() => screen.getByText(data.data[0].city));
      // Wait for barchart transition animation to complete (200 ms, set in
      // js/lib/chart_helpers/barchart.js)
      await delay(300);
    });

    it("renders a component with data loaded", () => {
      expect(component.asFragment()).toMatchSnapshot();
    });
  });

  describe("when data loading has an error", () => {
    const error = "you broke it";

    beforeEach(() => {
      console.error = jest.fn();
      DataLoader.loadRealtimeReportJSON.mockImplementation(() => {
        return Promise.reject(error);
      });
      component = render(
        <TopCities
          dataHrefBase="http://www.example.com/data/"
          refreshSeconds={30}
        />,
      );
    });

    it("renders a component in error state", () => {
      expect(component.asFragment()).toMatchSnapshot();
    });
  });
});
