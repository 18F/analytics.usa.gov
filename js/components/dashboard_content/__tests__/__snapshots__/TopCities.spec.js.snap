// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TopCities when data is loaded renders a component with data loaded 1`] = `
<DocumentFragment>
  <div
    class="grid-row"
  >
    <div
      class="chart__title display-flex card:grid-col-12 mobile-lg:grid-col-fill card:flex-justify-center mobile-lg:flex-justify-start card:padding-bottom-105 mobile-lg:padding-bottom-0"
    >
      <a
        aria-label="top-cities-realtime.csv"
        href="http://www.example.com/data//top-cities-realtime.csv"
      >
        <svg
          aria-hidden="true"
          class="usa-icon margin-bottom-neg-05 margin-left-05"
          focusable="false"
          role="img"
        >
          <use
            xlink:href="/assets/uswds/img/sprite.svg#file_present"
          />
        </svg>
      </a>
      <a
        href="/definitions#dimension_city"
      >
        <span
          class="usa-tooltip"
        >
          <span
            aria-describedby="tooltip-550000"
            class="usa-tooltip__trigger"
            data-position="top"
            tabindex="0"
          >
            Cities
          </span>
          <span
            aria-hidden="true"
            class="usa-tooltip__body"
            id="tooltip-550000"
            role="tooltip"
          >
            The city from which user activity originates. Location data may be affected by a user's VPN usage.
          </span>
        </span>
      </a>
    </div>
    <div
      class="card:grid-col-12 mobile-lg:grid-col-auto"
    >
      <div
        class="display-flex card:flex-justify-center mobile-lg:flex-justify-end"
      >
        <select
          aria-label="cities chart time filter"
          class="usa-select chart__filter margin-top-0"
        >
          <option
            value="top-cities-realtime"
          >
            30 Minutes
          </option>
          <option
            value="top-cities-yesterday"
          >
            Yesterday
          </option>
          <option
            value="top-cities-7-days"
          >
            7 Days
          </option>
          <option
            value="top-cities-30-days"
          >
            30 Days
          </option>
          <option
            value="top-cities-90-days"
          >
            90 Days
          </option>
          <option
            value="top-cities-current-year"
          >
            Current Year
          </option>
          <option
            value="top-cities-current-fiscal-year"
          >
            Current Fiscal Year
          </option>
          <option
            value="top-cities-previous-year"
          >
            Previous Year
          </option>
          <option
            value="top-cities-previous-fiscal-year"
          >
            Previous Fiscal Year
          </option>
        </select>
      </div>
    </div>
  </div>
  <div
    class="text--capitalize"
  >
    <figure
      id="chart_city"
    >
      <div
        class="data chart__bar-chart margin-top-2"
      >
        <div
          class="chart__bar-chart__item margin-bottom-2"
        >
          <div
            class="chart__bar-chart__item__label-wrapper grid-row"
          >
            <div
              class="chart__bar-chart__item__label dark-gray text--overflow-ellipsis grid-col-fill padding-right-1"
            >
              West Julianabury
            </div>
            <div
              class="chart__bar-chart__item__value grid-col-auto text--align-right"
            >
              19%
            </div>
          </div>
          <div
            class="chart__bar-chart__item__bar-wrapper grid-row"
          >
            <div
              class="chart__bar-chart__item__bar bg-palette-color-5 margin-top-0 grid-col-12"
              style="width: 19%;"
            />
          </div>
        </div>
        <div
          class="chart__bar-chart__item margin-bottom-2"
        >
          <div
            class="chart__bar-chart__item__label-wrapper grid-row"
          >
            <div
              class="chart__bar-chart__item__label dark-gray text--overflow-ellipsis grid-col-fill padding-right-1"
            >
              Palm Bay
            </div>
            <div
              class="chart__bar-chart__item__value grid-col-auto text--align-right"
            >
              18.3%
            </div>
          </div>
          <div
            class="chart__bar-chart__item__bar-wrapper grid-row"
          >
            <div
              class="chart__bar-chart__item__bar bg-palette-color-5 margin-top-0 grid-col-12"
              style="width: 18%;"
            />
          </div>
        </div>
        <div
          class="chart__bar-chart__item margin-bottom-2"
        >
          <div
            class="chart__bar-chart__item__label-wrapper grid-row"
          >
            <div
              class="chart__bar-chart__item__label dark-gray text--overflow-ellipsis grid-col-fill padding-right-1"
            >
              New Penelopeport
            </div>
            <div
              class="chart__bar-chart__item__value grid-col-auto text--align-right"
            >
              14.5%
            </div>
          </div>
          <div
            class="chart__bar-chart__item__bar-wrapper grid-row"
          >
            <div
              class="chart__bar-chart__item__bar bg-palette-color-5 margin-top-0 grid-col-12"
              style="width: 14%;"
            />
          </div>
        </div>
        <div
          class="chart__bar-chart__item margin-bottom-2"
        >
          <div
            class="chart__bar-chart__item__label-wrapper grid-row"
          >
            <div
              class="chart__bar-chart__item__label dark-gray text--overflow-ellipsis grid-col-fill padding-right-1"
            >
              New Alejandrachester
            </div>
            <div
              class="chart__bar-chart__item__value grid-col-auto text--align-right"
            >
              11.8%
            </div>
          </div>
          <div
            class="chart__bar-chart__item__bar-wrapper grid-row"
          >
            <div
              class="chart__bar-chart__item__bar bg-palette-color-5 margin-top-0 grid-col-12"
              style="width: 12%;"
            />
          </div>
        </div>
        <div
          class="chart__bar-chart__item margin-bottom-2"
        >
          <div
            class="chart__bar-chart__item__label-wrapper grid-row"
          >
            <div
              class="chart__bar-chart__item__label dark-gray text--overflow-ellipsis grid-col-fill padding-right-1"
            >
              Fort Althea
            </div>
            <div
              class="chart__bar-chart__item__value grid-col-auto text--align-right"
            >
              11.2%
            </div>
          </div>
          <div
            class="chart__bar-chart__item__bar-wrapper grid-row"
          >
            <div
              class="chart__bar-chart__item__bar bg-palette-color-5 margin-top-0 grid-col-12"
              style="width: 11%;"
            />
          </div>
        </div>
        <div
          class="chart__bar-chart__item margin-bottom-2"
        >
          <div
            class="chart__bar-chart__item__label-wrapper grid-row"
          >
            <div
              class="chart__bar-chart__item__label dark-gray text--overflow-ellipsis grid-col-fill padding-right-1"
            >
              Rodriguezshire
            </div>
            <div
              class="chart__bar-chart__item__value grid-col-auto text--align-right"
            >
              10.9%
            </div>
          </div>
          <div
            class="chart__bar-chart__item__bar-wrapper grid-row"
          >
            <div
              class="chart__bar-chart__item__bar bg-palette-color-5 margin-top-0 grid-col-12"
              style="width: 11%;"
            />
          </div>
        </div>
        <div
          class="chart__bar-chart__item margin-bottom-2"
        >
          <div
            class="chart__bar-chart__item__label-wrapper grid-row"
          >
            <div
              class="chart__bar-chart__item__label dark-gray text--overflow-ellipsis grid-col-fill padding-right-1"
            >
              Felixtown
            </div>
            <div
              class="chart__bar-chart__item__value grid-col-auto text--align-right"
            >
              5%
            </div>
          </div>
          <div
            class="chart__bar-chart__item__bar-wrapper grid-row"
          >
            <div
              class="chart__bar-chart__item__bar bg-palette-color-5 margin-top-0 grid-col-12"
              style="width: 5%;"
            />
          </div>
        </div>
        <div
          class="chart__bar-chart__item margin-bottom-2"
        >
          <div
            class="chart__bar-chart__item__label-wrapper grid-row"
          >
            <div
              class="chart__bar-chart__item__label dark-gray text--overflow-ellipsis grid-col-fill padding-right-1"
            >
              Alexanebury
            </div>
            <div
              class="chart__bar-chart__item__value grid-col-auto text--align-right"
            >
              4.6%
            </div>
          </div>
          <div
            class="chart__bar-chart__item__bar-wrapper grid-row"
          >
            <div
              class="chart__bar-chart__item__bar bg-palette-color-5 margin-top-0 grid-col-12"
              style="width: 5%;"
            />
          </div>
        </div>
        <div
          class="chart__bar-chart__item margin-bottom-2"
        >
          <div
            class="chart__bar-chart__item__label-wrapper grid-row"
          >
            <div
              class="chart__bar-chart__item__label dark-gray text--overflow-ellipsis grid-col-fill padding-right-1"
            >
              Erynfort
            </div>
            <div
              class="chart__bar-chart__item__value grid-col-auto text--align-right"
            >
              3.7%
            </div>
          </div>
          <div
            class="chart__bar-chart__item__bar-wrapper grid-row"
          >
            <div
              class="chart__bar-chart__item__bar bg-palette-color-5 margin-top-0 grid-col-12"
              style="width: 4%;"
            />
          </div>
        </div>
        <div
          class="chart__bar-chart__item margin-bottom-2"
        >
          <div
            class="chart__bar-chart__item__label-wrapper grid-row"
          >
            <div
              class="chart__bar-chart__item__label dark-gray text--overflow-ellipsis grid-col-fill padding-right-1"
            >
              Lake Taya
            </div>
            <div
              class="chart__bar-chart__item__value grid-col-auto text--align-right"
            >
              1.2%
            </div>
          </div>
          <div
            class="chart__bar-chart__item__bar-wrapper grid-row"
          >
            <div
              class="chart__bar-chart__item__bar bg-palette-color-5 margin-top-0 grid-col-12"
              style="width: 1%;"
            />
          </div>
        </div>
      </div>
    </figure>
  </div>
</DocumentFragment>
`;

exports[`TopCities when data is not loaded renders a component in loading state 1`] = `
<DocumentFragment>
  <div
    class="grid-row"
  >
    <div
      class="chart__title display-flex card:grid-col-12 mobile-lg:grid-col-fill card:flex-justify-center mobile-lg:flex-justify-start card:padding-bottom-105 mobile-lg:padding-bottom-0"
    >
      <a
        aria-label="top-cities-realtime.csv"
        href="http://www.example.com/data//top-cities-realtime.csv"
      >
        <svg
          aria-hidden="true"
          class="usa-icon margin-bottom-neg-05 margin-left-05"
          focusable="false"
          role="img"
        >
          <use
            xlink:href="/assets/uswds/img/sprite.svg#file_present"
          />
        </svg>
      </a>
      <a
        href="/definitions#dimension_city"
      >
        <span
          class="usa-tooltip"
        >
          <span
            aria-describedby="tooltip-550000"
            class="usa-tooltip__trigger"
            data-position="top"
            tabindex="0"
          >
            Cities
          </span>
          <span
            aria-hidden="true"
            class="usa-tooltip__body"
            id="tooltip-550000"
            role="tooltip"
          >
            The city from which user activity originates. Location data may be affected by a user's VPN usage.
          </span>
        </span>
      </a>
    </div>
    <div
      class="card:grid-col-12 mobile-lg:grid-col-auto"
    >
      <div
        class="display-flex card:flex-justify-center mobile-lg:flex-justify-end"
      >
        <select
          aria-label="cities chart time filter"
          class="usa-select chart__filter margin-top-0"
        >
          <option
            value="top-cities-realtime"
          >
            30 Minutes
          </option>
          <option
            value="top-cities-yesterday"
          >
            Yesterday
          </option>
          <option
            value="top-cities-7-days"
          >
            7 Days
          </option>
          <option
            value="top-cities-30-days"
          >
            30 Days
          </option>
          <option
            value="top-cities-90-days"
          >
            90 Days
          </option>
          <option
            value="top-cities-current-year"
          >
            Current Year
          </option>
          <option
            value="top-cities-current-fiscal-year"
          >
            Current Fiscal Year
          </option>
          <option
            value="top-cities-previous-year"
          >
            Previous Year
          </option>
          <option
            value="top-cities-previous-fiscal-year"
          >
            Previous Fiscal Year
          </option>
        </select>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`TopCities when data loading has an error renders a component in error state 1`] = `
<DocumentFragment>
  <div
    class="grid-row"
  >
    <div
      class="chart__title display-flex card:grid-col-12 mobile-lg:grid-col-fill card:flex-justify-center mobile-lg:flex-justify-start card:padding-bottom-105 mobile-lg:padding-bottom-0"
    >
      <a
        aria-label="top-cities-realtime.csv"
        href="http://www.example.com/data//top-cities-realtime.csv"
      >
        <svg
          aria-hidden="true"
          class="usa-icon margin-bottom-neg-05 margin-left-05"
          focusable="false"
          role="img"
        >
          <use
            xlink:href="/assets/uswds/img/sprite.svg#file_present"
          />
        </svg>
      </a>
      <a
        href="/definitions#dimension_city"
      >
        <span
          class="usa-tooltip"
        >
          <span
            aria-describedby="tooltip-550000"
            class="usa-tooltip__trigger"
            data-position="top"
            tabindex="0"
          >
            Cities
          </span>
          <span
            aria-hidden="true"
            class="usa-tooltip__body"
            id="tooltip-550000"
            role="tooltip"
          >
            The city from which user activity originates. Location data may be affected by a user's VPN usage.
          </span>
        </span>
      </a>
    </div>
    <div
      class="card:grid-col-12 mobile-lg:grid-col-auto"
    >
      <div
        class="display-flex card:flex-justify-center mobile-lg:flex-justify-end"
      >
        <select
          aria-label="cities chart time filter"
          class="usa-select chart__filter margin-top-0"
        >
          <option
            value="top-cities-realtime"
          >
            30 Minutes
          </option>
          <option
            value="top-cities-yesterday"
          >
            Yesterday
          </option>
          <option
            value="top-cities-7-days"
          >
            7 Days
          </option>
          <option
            value="top-cities-30-days"
          >
            30 Days
          </option>
          <option
            value="top-cities-90-days"
          >
            90 Days
          </option>
          <option
            value="top-cities-current-year"
          >
            Current Year
          </option>
          <option
            value="top-cities-current-fiscal-year"
          >
            Current Fiscal Year
          </option>
          <option
            value="top-cities-previous-year"
          >
            Previous Year
          </option>
          <option
            value="top-cities-previous-fiscal-year"
          >
            Previous Fiscal Year
          </option>
        </select>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
